src/gui/plugins/align_tool/AlignTool.cc

```diff
@@ -0,0 +1,647 @@
+{
+  std::lock_guard<std::mutex> lock(this->dataPtr->mutex);
+  std::string newAxis = _axis.toStdString();
+  std::transform(newAxis.begin(), newAxis.end(), newAxis.begin(), ::tolower);
+
+  if (newAxis == "x")
+  {
+    this->dataPtr->axis = AlignAxis::ALIGN_X;
+  }
+  else if (newAxis == "y")
+  {
+    this->dataPtr->axis = AlignAxis::ALIGN_Y;
+  }
+  else if (newAxis == "z")
+  {
+    this->dataPtr->axis = AlignAxis::ALIGN_Z;
+  }
+  else
+  {
+    ignwarn << "Invalid align axis string: " << newAxis << "\n";
+    ignwarn << "The valid options are:\n";
+    ignwarn << " - X\n";
+    ignwarn << " - Y\n";
+    ignwarn << " - Z\n";
+  }
+}
+
+/////////////////////////////////////////////////
+void AlignTool::OnAlignTarget(const QString &_target)
+{
+  std::lock_guard<std::mutex> lock(this->dataPtr->mutex);
+  std::string newTarget = _target.toStdString();
+  std::transform(newTarget.begin(), newTarget.end(),
+                 newTarget.begin(), ::tolower);
+
+  if (newTarget == "first")
+  {
+    this->dataPtr->first = true;
+  }
+  else if (newTarget == "last")
+  {
+    this->dataPtr->first = false;
+  }
+  else
+  {
+    ignwarn << "Invalid align target string: " << newTarget << "\n";
+    ignwarn << "The valid options are:\n";
+    ignwarn << " - first\n";
+    ignwarn << " - last\n";
+  }
+}
+
+/////////////////////////////////////////////////
+void AlignTool::OnHoveredEntered()
+{
+  std::lock_guard<std::mutex> lock(this->dataPtr->mutex);
+  this->AddState(AlignState::HOVER);
+}
+
+/////////////////////////////////////////////////
+void AlignTool::OnHoveredExited()
+{
+  std::lock_guard<std::mutex> lock(this->dataPtr->mutex);
+  // Always reset after an exit, if an align call is made, it
+  // will have updated the prev positions of the entities so a
+  // reset after align will result in no change
+  this->AddState(AlignState::RESET);
+}
+
+/////////////////////////////////////////////////
+void AlignTool::OnAlign()
+{
+  this->AddState(AlignState::ALIGN);
+}
+
+/////////////////////////////////////////////////
+void AlignTool::AddState(const AlignState &_state)
+{
+  this->dataPtr->states.push(_state);
+}
+
+////////////////////////////////////////////////
+void AlignTool::MakeTransparent(const rendering::NodePtr &_node)
+{
+  if (!_node)
+    return;
+
+  for (auto n = 0u; n < _node->ChildCount(); ++n)
+  {
+    auto child = _node->ChildByIndex(n);
+    this->MakeTransparent(child);
+  }
+
+  auto vis = std::dynamic_pointer_cast<rendering::Visual>(_node);
+  if (nullptr == vis)
+    return;
+
+  // Visual material
+  auto visMat = vis->Material();
+  if (nullptr != visMat)
+  {
+    // If the entity isn't already transparent, make it transparent
+    if (this->dataPtr->originalTransparency.find(vis->Name()) ==
+        this->dataPtr->originalTransparency.end())
+    {
+      this->dataPtr->originalTransparency[vis->Name()] = visMat->Transparency();
+      visMat->SetTransparency(visMat->Transparency() + 0.5);
```

Comment:
if a visual's transparency's setting is > 0.5, the logic here may make the visual fully transparent. What do you think about multiplying the effect instead of addition? something like:

```
visMat->SetTransparency(1.0 - ((1.0 - visMat->Transparency()) * 0.5));
```

e.g. for a visual with 0.6 transparency, the new transparency value would be: 1.0 - ((1.0 - 0.6)*0.5) = 0.8

Reply:
Resolved in [afe5d75b](https://github.com/ignitionrobotics/ign-gazebo/pull/59/commits/afe5d75b8f2ee1b8b38247903c54b5d596f02760)